core/src/main/java/org/apache/carbondata/core/scan/result/vector/impl/directread/ColumnarVectorWrapperDirectWithDeleteDeltaAndInvertedIndex.java

```diff
@@ -0,0 +1,258 @@
+import org.apache.carbondata.core.metadata.datatype.DecimalType;
+import org.apache.carbondata.core.scan.result.vector.CarbonColumnVector;
+import org.apache.carbondata.core.scan.result.vector.impl.CarbonColumnVectorImpl;
+
+/**
+ * Column vector for column pages which has delete delta and inverted index, so it uses delta biset
+ * to filter out data and use inverted index before filling to actual vector
+ */
+class ColumnarVectorWrapperDirectWithDeleteDeltaAndInvertedIndex
+    extends AbstractCarbonColumnarVector implements ConvertableVector {
+
+  private BitSet deletedRows;
+
+  private int[] invertedIndex;
+
+  private CarbonColumnVectorImpl carbonColumnVector;
+
+  private CarbonColumnVector columnVector;
+
+  private int precision;
+
+  private BitSet nullBits;
+
+  private boolean isnullBitsExists;
+
+  public ColumnarVectorWrapperDirectWithDeleteDeltaAndInvertedIndex(
+      CarbonColumnVector vectorWrapper, BitSet deletedRows, int[] invertedIndex, BitSet nullBits,
+      boolean isnullBitsExists) {
+    this.deletedRows = deletedRows;
+    this.invertedIndex = invertedIndex;
+    carbonColumnVector = new CarbonColumnVectorImpl(invertedIndex.length, vectorWrapper.getType());
+    this.columnVector = vectorWrapper;
+    this.nullBits = nullBits;
+    this.isnullBitsExists = isnullBitsExists;
+  }
+
+  @Override
+  public void putBoolean(int rowId, boolean value) {
+    carbonColumnVector.putBoolean(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putFloat(int rowId, float value) {
+    carbonColumnVector.putFloat(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putShort(int rowId, short value) {
+    carbonColumnVector.putShort(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putInt(int rowId, int value) {
+    carbonColumnVector.putInt(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putLong(int rowId, long value) {
+    carbonColumnVector.putLong(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putDecimal(int rowId, BigDecimal value, int precision) {
+    this.precision = precision;
+    carbonColumnVector.putDecimal(invertedIndex[rowId], value, precision);
+  }
+
+  @Override
+  public void putDouble(int rowId, double value) {
+    carbonColumnVector.putDouble(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putByteArray(int rowId, byte[] value) {
+    carbonColumnVector.putByteArray(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putByteArray(int rowId, int offset, int length, byte[] value) {
+    carbonColumnVector.putByteArray(invertedIndex[rowId], offset, length, value);
+  }
+
+  @Override
+  public void putByte(int rowId, byte value) {
+    carbonColumnVector.putByte(invertedIndex[rowId], value);
+  }
+
+  @Override
+  public void putNull(int rowId) {
+    if (isnullBitsExists) {
+      nullBits.set(rowId);
+    } else {
+      nullBits.set(invertedIndex[rowId]);
```

Comment:
please add a comment to explain this if-else logic...flag `isnullBitsExists` is to distinguish between measure and dimension. When the flag is true that means it is a measure and for measures we do not have inverted index while the else case is for dimensions  for which inverted index exists. Is my understanding correct?

Reply:
No, there is no differentiation between measure and dimension at this level.  we are having nullbitset even for dimensions also.  But some dimension columns still don't have nullbitset. So if null bitset does not exist then it should not inverted index while setting the null. Same will be updated in the comment.